### PR TITLE
11: Pass for Twitter added.

### PR DESCRIPTION
**Pull request by popprem**
Pass for twitter added. Todo: PsTwitterTest has to be implemented or XeTwitterLink & XeTwitterLinkTest have to be implemented which will test PsTwitter too.

### CODE REVIEW

**Comment by davvd** [comment]
@popprem Thanks, I will find someone to review this PR soon

**Comment by davvd** [comment]
@longtimeago please review,thanks

File: pom.xml [hunk 1]
```diff
@@ -136,6 +136,11 @@
             <optional>true</optional>
         </dependency>
         <dependency>
+		    <groupId>commons-codec</groupId>
```
Comment (longtimeago) [inline pom.xml]: @popprem we are not allowed to use external dependency with compile scope. We promised that to our clients.

File: src/main/java/org/takes/facets/auth/social/EncodingException.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+package org.takes.facets.auth.social;
+
+/**
+ * Encoding exception.
+ *
+ * @author Prasath Premkumar (popprem@gmail.com)
+ * @version $Id$
+ * @since
+ */
+public final class EncodingException extends RuntimeException {
```
Comment (longtimeago) [inline src/main/java/org/takes/facets/auth/social/EncodingException.java]: @popprem don't see any reason to add this custom exception.  Should the client handle the exception somehow? I think no.

File: src/main/java/org/takes/facets/auth/social/PsTwitter.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+            final String consumersecret) {
+        try {
+            final String authkey = URLEncoder.encode(consumerkey, ENCODING)
+                    + ":" + URLEncoder.encode(consumersecret, ENCODING);
+            final byte[] encodedBytes = Base64.encodeBase64(
+                    authkey.getBytes()
+            );
+            return new String(encodedBytes);
+        } catch (final UnsupportedEncodingException  ex) {
+            throw new EncodingException(ex);
```
Comment (longtimeago) [inline src/main/java/org/takes/facets/auth/social/PsTwitter.java]: Just throw ```IllegalArgumentException``` with proper cause.

File: src/main/java/org/takes/facets/auth/social/PsTwitter.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+ * @since
+ * @checkstyle MultipleStringLiteralsCheck (500 lines)
+ */
+@EqualsAndHashCode(of = { "app", "key" })
+public final class PsTwitter implements Pass {
+
+    /**
+     * UTF-8 encoding.
+     */
+    private static final String ENCODING = "UTF-8";
```
Comment (longtimeago) [inline src/main/java/org/takes/facets/auth/social/PsTwitter.java]: @popprem you would not need this when switched on ```DatatypeConverter```

File: src/main/java/org/takes/facets/auth/social/PsTwitter.java [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        );
+    }
+
+    /**
+     * Encode consumerKey and consumerSecret and prepare authorization key.
+     * @param consumerkey Twitter consumerKey
+     * @param consumersecret Twitter consumerSecret
+     * @return String authorization key
+     */
+    private static String prepareAuthKey(final String consumerkey,
```
Comment (longtimeago) [inline src/main/java/org/takes/facets/auth/social/PsTwitter.java]: don't think we would need this method at all as soon as migrated on DatatypeConverter

File: src/main/java/org/takes/facets/auth/social/PsTwitter.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+            .toString();
+        return new JdkRequest(uri)
+            .method("POST")
+                .header(
+                        "Content-Type",
+                        "application/x-www-form-urlencoded;charset=UTF-8"
+            )
+                .header(
+                        "Authorization",
+                        "Basic ".concat(
```
Comment (longtimeago) [inline src/main/java/org/takes/facets/auth/social/PsTwitter.java]: it's more natural to use ```String.format()``` here

**Comment by longtimeago** [comment]
@popprem thanks for the PR. See several remarks above

**Comment by popprem** [comment]
@longtimeago Thanks for the review, have addressed the issues. Please check.

**Comment by longtimeago** [comment]
@popprem Now we are good to go, thank you!

**Comment by longtimeago** [comment]
@rultor merge please

**Comment by longtimeago** [comment]
@popprem looks like I've missed to note one important thing ... no tests ?? Please, add at least one. Yegor will not leave out this PR without test

**Comment by popprem** [comment]
@longtimeago Yes, I have commented the same on the pull request. Test for PsTwitter is not straightforward ,should be done with oauth mocks etc. Need advise on the same.

**Comment by longtimeago** [comment]
@popprem yep ... I see the comment. Add at least todo in the code. Other developer will implement it

**Comment by popprem** [comment]
@longtimeago Test added with todo comment, please check.

**Comment by longtimeago** [comment]
@popprem Todo must be formatted as puzzle (see examples https://github.com/teamed/pdd)
``` @todo #11:30min  What actually to be done```
Description should be meaningful, more that 20 words. New task will be generated by this todo with description you provided. So, please, put there as much details as possible to avoid questions from next developer. Imagine, you are the developer and will get the task, generated by this todo. What would you like to be mentioned there?

**Comment by popprem** [comment]
@longtimeago Test description updated with proper todo format. Please verify.

**Comment by longtimeago** [comment]
@popprem now we are good to go, thanks!

**Comment by longtimeago** [comment]
@rultor merge please

File: src/main/java/org/takes/facets/auth/social/PsTwitter.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        return new JdkRequest(uri)
+            .method("POST")
+                .header(
+                        "Content-Type",
+                        "application/x-www-form-urlencoded;charset=UTF-8"
+            )
+                .header(
+                        "Authorization",
+                        String.format(
+                                "Basic %s", DatatypeConverter
```
Comment (yegor256) [inline src/main/java/org/takes/facets/auth/social/PsTwitter.java]: why eight indentation spaces here?

**Comment by yegor256** [comment]
@popprem there is something crazy going on with indentation. I can't read the code. please, fix it

**Comment by longtimeago** [comment]
@yegor256 Bug in qulice?

**Comment by yegor256** [comment]
@longtimeago well, I don't think so... the indentation used in this branch probably satisfies qulice, but it doesn't have any logic behind. this is what happens when you trying to cut corners too much :)

**Comment by popprem** [comment]
@yegor256 @longtimeago indentation is according to qulice. If i do any formatting, qulice is not happy and fails the build. Pls advice.

**Comment by yegor256** [comment]
@popprem you should satisfy both - qulice and us. Just qulice is not enough. The code should be readable by humans too

**Comment by popprem** [comment]
@yegor256 Updated indentation to make it more readable, please check if this is ok with you.

**Comment by yegor256** [comment]
@popprem I still don't understand why do you use 8 indentation spaces sometimes and sometimes 4. what is the message here? what are you trying to say by 8 spaces? also, read this: http://www.yegor256.com/2014/10/23/paired-brackets-notation.html

**Comment by popprem** [comment]
@yegor256 Looks cleaner now, please verify.

**Comment by yegor256** [comment]
@rultor merge

**Comment by longtimeago** [comment]
@popprem please, merge from master

**Comment by popprem** [comment]
@longtimeago Have merged from master. Please try merging.

**Comment by longtimeago** [comment]
@rultor please merge

**Comment by popprem** [comment]
@yegor256 Can this be merged?

**Comment by yegor256** [comment]
@rultor merge

**Comment by popprem** [comment]
@davvd @yegor256 Can this be closed please?

**Comment by yegor256** [comment]
@popprem it is closed already

**Comment by davvd** [comment]
@elenavolokhova review this ticket please, for compliance with our [quality rules](http://www.teamed.io/qa.html)

**Comment by davvd** [comment]
@rultor deploy pls

**Comment by elenavolokhova** [comment]
@longtimeago 
Please fix description of this PR. Pull request description should explain the solution proposed and contains a link to the original ticket it is related to.

**Comment by popprem** [comment]
@elenavolokhova Added link to the original ticket.

**Comment by elenavolokhova** [comment]
@popprem Thank you!

**Comment by elenavolokhova** [comment]
@davvd Quality is acceptable

**Comment by davvd** [comment]
> @davvd Quality is acceptable

@elenavolokhova thanks for the QA review, we'll work better next time

**Comment by davvd** [comment]
> @davvd Quality is acceptable

@elenavolokhova thank you, let's try to make it "good" next time

**Comment by davvd** [comment]
> @davvd Quality is acceptable

@elenavolokhova thank you, let's try to make it "good" next time

**Comment by davvd** [comment]
> @davvd Quality is acceptable

@elenavolokhova thanks for the review, we'll try better next time

**Comment by davvd** [comment]
> @davvd Quality is acceptable

@elenavolokhova thanks for the review, we'll try better next time

**Comment by davvd** [comment]
> @davvd Quality is acceptable

@elenavolokhova thanks a lot, next time everybody should try to make it better

**Comment by davvd** [comment]
@longtimeago I added 10 mins to @elenavolokhova (for QA review) in transaction 56850215... Thanks a lot, I just topped your account for **26 mins**, transaction ID `56850223`, total time was 332 hours and 21 mins... you're getting extra minutes here (c=11)... +26 to your rating, your total score is [+4581](http://www.netbout.com/b/35024?open=rating)

**Pull request merged**